### PR TITLE
[aot] Add support for making direct cross-assembly calls in llvm mode.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by lateralusX** [comment]
Testing the direct-external-calls a little, couple of things I have seen so far:

- Link error, System.dll-llvm.obj : error LNK2001: unresolved external symbol aot_mscorlib_System_System_dot_AsyncCallback__BeginInvoke_clsc_IAsyncResult__this_clsc_IAsyncResult_clsd_AsyncCallback_obj, looks like this method is still emitted expecting to be called through PLT, but caller try to call it using mangled direct callable name and that symbol doesnt exist in mscorlib object files. Do we need to exclude more wrappers in method_is_externally_callable to exclude these as well?

- Comparing against the llvmonly, it allows method->wrapper_type == MONO_WRAPPER_ALLOC to be direct callable, but not done in none llvmonly, any reasons for that?

- What is preventing us from doing direct calls for method->is_inflated? I see that we can't even do that for calls within assembly, so I guess it's not doable since we need to setup class only when first called to preserve semantics of generics?

- No direct calls for string ctors, what is preventing us from doing that?

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -13471,6 +13572,11 @@ mono_compile_assembly (MonoAssembly *ass, guint32 opts, const char *aot_options,
 
 	mono_aot_parse_options (aot_options, &acfg->aot_opts);
 
+	if (acfg->aot_opts.direct_extern_calls && !(acfg->aot_opts.llvm && acfg->aot_opts.static_link)) {
+		fprintf (stderr, "The 'direct-extern-calls' option requires the 'llvm' and 'static' options.\n");
```
Comment (lewurm) [inline mono/mini/aot-compiler.c]: `aot_printerrf`?

File: mono/mini/mini-llvm.c [hunk 2]
```diff
@@ -2107,6 +2109,49 @@ get_callee (EmitContext *ctx, LLVMTypeRef llvm_sig, MonoJumpInfoType type, gcons
...
+
+				callee_name = mono_aot_get_mangled_method_name (cmethod);
+
+				callee = (LLVMValueRef)g_hash_table_lookup (ctx->module->direct_callables, callee_name);
+				if (!callee) {
+					callee = LLVMAddFunction (ctx->lmodule, callee_name, llvm_sig);
+
+					LLVMSetLinkage (callee, LLVMExternalLinkage);
+
+					g_hash_table_insert (ctx->module->direct_callables, (char*)callee_name, callee);
```
Comment (lewurm) [inline mono/mini/mini-llvm.c]: `(char*)` cast not needed

**Comment by lateralusX** [comment]
Seeing a ~10% perf drop on test sample when enabling direct-extern-calls. No calls are done against init_method or to the per assembly init wrappers, so looks like the init checks are working as expected. This sample will have methods that will fail LLVM compile, some of them might end up in hot paths.

**Pull request merged**